clang/include/clang/CConv/ConstraintsGraph.h

```diff
@@ -12,125 +12,269 @@
+  // be able to retrieve them from the graph.
+  std::set<ConstAtom*> &getAllConstAtoms();
+
+protected:
+  // Add vertex is overridden to save const atoms as they are added to the graph
+  // so that getAllConstAtoms can efficiently retrieve them.
+  NodeType *addVertex(Atom *A) override;
+private:
+  std::set<ConstAtom*> AllConstAtoms;
 };
 
-class GraphVizOutputGraph
-    : public BaseGraph<
-        boost::adjacency_list<boost::vecS, boost::vecS, boost::bidirectionalS,
-                                Atom *, EdgeProperties *>> {
-public:
-  typedef boost::adjacency_list<boost::vecS, boost::vecS, boost::bidirectionalS,
-                                Atom *, EdgeProperties *>
-      DirectedGraphType;
 
-  void mergeConstraintGraph(const ConstraintsGraph &Graph, EdgeType EdgeType);
+// Below this point we define a graph class specialized for generating the
+// graphviz output for the combine checked and ptype graphs.
+
+// Once we combine the graphs, we need to know if an edge came from the ptype
+// or the checked constraint graph. We also combine edges to render a pair of
+// edges pointing in different directions between the same pair of nodes as a
+// single bidirectional edge.
+class GraphVizEdge :
+    public llvm::DGEdge<DataNode<Atom*, GraphVizEdge>, GraphVizEdge> {
+public:
+  enum EdgeKind { EK_Checked, EK_Ptype };
+  explicit GraphVizEdge(DataNode<Atom *, GraphVizEdge> &Node, EdgeKind Kind)
+      : DGEdge(Node), Kind(Kind), IsBidirectional(false) {}
+  GraphVizEdge(const GraphVizEdge &E)
+      : DGEdge(E), Kind(E.Kind), IsBidirectional(E.IsBidirectional) {}
 
-  // Dump the graph to stdout in a dot format.
-  void dumpCGDot(const std::string& GraphDotFile);
+  EdgeKind Kind;
+  bool IsBidirectional;
+};
 
+// The graph subclass for graphviz output uses the specialized edge class to
+// hold the extra information required. It also provides the methods for
+// merging the checked and ptype graphs as well as a static function to do the
+// actual conversion and output to file.
+class GraphVizOutputGraph : public DataGraph<Atom*, GraphVizEdge> {
+public:
+  // Merge the provided graphs and dump the graphviz visualization to the given
+  // file name. The first graph argument is the checked graph while the second
+  // is the pointer type graph.
   static void dumpConstraintGraphs(const std::string &GraphDotFile,
                                    const ConstraintsGraph &Chk,
                                    const ConstraintsGraph &Pty);
 
+  // These maps are used because the graphviz utility provided by llvm does not
+  // give an easy way to differentiate between multiple edges between the same
+  // pair of nodes. When there is both a checked an a ptype edge between two
+  // nodes, these maps ensure that each is output exactly once and has the
+  // correct color when it is output.
+  mutable std::set<std::pair<Atom *,Atom *>> DoneChecked;
+  mutable std::set<std::pair<Atom *,Atom *>> DonePtyp;
```

Comment:
Why do you need to explicitly indicate `mutable` here? I've never seen that (!).

Reply:
It's a bit of a hack. The graphviz output utility gives you a `const` references to the graph your working with, but I needed to mutate these maps. `mutable` lets you mutate fields on `const` objects if you really want to.